src/descriptor/mod.rs

```diff
@@ -619,69 +447,161 @@ impl Descriptor<DescriptorPublicKey> {
 where
     Pk: MiniscriptKey + str::FromStr,
     Pk::Hash: str::FromStr,
     <Pk as FromStr>::Err: ToString,
     <<Pk as MiniscriptKey>::Hash as FromStr>::Err: ToString,
 {
     /// Parse an expression tree into a descriptor
-    fn from_tree(top: &expression::Tree) -> Result<Descriptor<Pk>, Error> {
+    fn from_tree(top: &expression::Tree) -> Result<PreTaprootDescriptor<Pk>, Error> {
         Ok(match (top.name, top.args.len() as u32) {
-            ("pkh", 1) => Descriptor::Pkh(Pkh::from_tree(top)?),
-            ("wpkh", 1) => Descriptor::Wpkh(Wpkh::from_tree(top)?),
-            ("sh", 1) => Descriptor::Sh(Sh::from_tree(top)?),
-            ("wsh", 1) => Descriptor::Wsh(Wsh::from_tree(top)?),
-            ("tr", _) => Descriptor::Tr(Tr::from_tree(top)?),
-            _ => Descriptor::Bare(Bare::from_tree(top)?),
+            ("pkh", 1) => PreTaprootDescriptor::Pkh(Pkh::from_tree(top)?),
+            ("wpkh", 1) => PreTaprootDescriptor::Wpkh(Wpkh::from_tree(top)?),
+            ("sh", 1) => PreTaprootDescriptor::Sh(Sh::from_tree(top)?),
+            ("wsh", 1) => PreTaprootDescriptor::Wsh(Wsh::from_tree(top)?),
+            _ => PreTaprootDescriptor::Bare(Bare::from_tree(top)?),
         })
     }
 }
 
-impl<Pk> FromStr for Descriptor<Pk>
-where
-    Pk: MiniscriptKey + str::FromStr,
-    Pk::Hash: str::FromStr,
-    <Pk as FromStr>::Err: ToString,
-    <<Pk as MiniscriptKey>::Hash as FromStr>::Err: ToString,
-{
-    type Err = Error;
+serde_string_impl_pk!(PreTaprootDescriptor, "a pre-taproot script descriptor");
+pub(crate) mod pretaproot {
 
-    fn from_str(s: &str) -> Result<Descriptor<Pk>, Error> {
-        let desc_str = verify_checksum(s)?;
-        let top = expression::Tree::from_str(desc_str)?;
-        expression::FromTree::from_tree(&top)
-    }
-}
+    use bitcoin::{self, blockdata::witness::Witness, Script};
 
-impl<Pk: MiniscriptKey> fmt::Debug for Descriptor<Pk> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            Descriptor::Bare(ref sub) => write!(f, "{:?}", sub),
-            Descriptor::Pkh(ref pkh) => write!(f, "{:?}", pkh),
-            Descriptor::Wpkh(ref wpkh) => write!(f, "{:?}", wpkh),
-            Descriptor::Sh(ref sub) => write!(f, "{:?}", sub),
-            Descriptor::Wsh(ref sub) => write!(f, "{:?}", sub),
-            Descriptor::Tr(ref tr) => write!(f, "{:?}", tr),
+    use super::{Pkh, PreTaprootDescriptor, Sh, Wpkh, Wsh};
+    use {DescriptorTrait, Error, MiniscriptKey, Satisfier, ToPublicKey};
+
+    /// A general trait for Pre taproot bitcoin descriptor.
+    /// Similar to [`DescriptorTrait`], but `explicit_script` and `script_code` methods cannot fail
+    pub trait PreTaprootDescriptorTrait<Pk: MiniscriptKey>: DescriptorTrait<Pk> {
+        /// Same as [`DescriptorTrait::sanity_check`]
+        fn sanity_check(&self) -> Result<(), Error> {
+            <Self as DescriptorTrait<Pk>>::sanity_check(&self)
         }
```

Comment:
I think these pass-through implementaitons aren't needed. Because `PreTaprootDescriptorTrait` implies `DescriptorTrait`, anything that implements the former is guaranteed to implement the latter.

I think that users might be forced to import both traits in order to use their methods, but I think this is a reasonable cost to charge them, in exchange for greatly reducing the amount of duplicated code in this crate.

Reply:
I see. This is much better